pkg/orchestrator/scanestimationwatcher/watcher.go

```diff
@@ -0,0 +1,574 @@
+		assetErrs = append(assetErrs, err)
+	}
+	numOfErrs := len(assetErrs)
+
+	if numOfErrs > 0 {
+		return fmt.Errorf("failed to create %d AssetScanEstimation(s) for ScanEstimation. ScanEstimationID=%s: %w", numOfErrs, *scanEstimation.Id, assetErrs[0])
+	}
+
+	if scanEstimation.Summary == nil {
+		scanEstimation.Summary = &models.ScanEstimationSummary{}
+	}
+	scanEstimation.Summary.JobsLeftToRun = utils.PointerTo(numOfAssets)
+
+	return nil
+}
+
+func (w *Watcher) newAssetScanEstimationFromScanEstimation(scanEstimation *models.ScanEstimation, assetID string) (models.AssetScanEstimation, error) {
+	if scanEstimation == nil {
+		return models.AssetScanEstimation{}, fmt.Errorf("empty scan estimation")
+	}
+
+	if scanEstimation.ScanTemplate == nil {
+		return models.AssetScanEstimation{}, fmt.Errorf("empty scan template")
+	}
+
+	return models.AssetScanEstimation{
+		Asset: &models.AssetRelationship{
+			Id: assetID,
+		},
+		AssetScanTemplate: scanEstimation.ScanTemplate.AssetScanTemplate,
+		ScanEstimation: &models.ScanEstimationRelationship{
+			Id: scanEstimation.Id,
+		},
+		State: &models.AssetScanEstimationState{
+			LastTransitionTime: utils.PointerTo(time.Now()),
+			State:              utils.PointerTo(models.AssetScanEstimationStateStatePending),
+		},
+	}, nil
+}
+
+func (w *Watcher) createAssetScanEstimationForAsset(ctx context.Context, scanEstimation *models.ScanEstimation, assetID string) error {
+	logger := log.GetLoggerFromContextOrDiscard(ctx)
+
+	assetScanEstimationData, err := w.newAssetScanEstimationFromScanEstimation(scanEstimation, assetID)
+	if err != nil {
+		return fmt.Errorf("failed to generate new AssetScanEstimation for ScanEstimation. ScanEstimationID=%s, AssetID=%s: %w", *scanEstimation.Id, assetID, err)
+	}
+
+	ret, err := w.backend.PostAssetScanEstimation(ctx, assetScanEstimationData)
+	if err != nil {
+		var conErr backendclient.AssetScanEstimationConflictError
+		if errors.As(err, &conErr) {
+			assetScanEstimationID := *conErr.ConflictingAssetScanEstimation.Id
+			logger.WithField("AssetScanEstimationID", assetScanEstimationID).Debug("AssetScanEstimation already exist.")
+			return nil
+		}
+		return fmt.Errorf("failed to post AssetScanEstimation to backend API: %w", err)
+	}
+
+	if scanEstimation.AssetScanEstimations == nil {
+		scanEstimation.AssetScanEstimations = &[]models.AssetScanEstimationRelationship{}
+	}
+	*scanEstimation.AssetScanEstimations = append(*scanEstimation.AssetScanEstimations, models.AssetScanEstimationRelationship{Id: ret.Id})
+
+	return nil
+}
+
+func updateScanEstimationSummaryFromAssetScanEstimation(scanEstimation *models.ScanEstimation, result models.AssetScanEstimation) error {
+	state, ok := result.GetState()
+	if !ok {
+		return fmt.Errorf("state must not be nil for AssetScan. AssetScanID=%s", *result.Id)
+	}
+
+	s := scanEstimation.Summary
+
+	switch state {
+	case models.AssetScanEstimationStateStatePending:
+		s.JobsLeftToRun = utils.PointerTo(*s.JobsLeftToRun + 1)
```

Comment:
do you assume to call this function on every asset scan once? Because if this reconcile function is called several times for the same asset scan estimation and the same scan estimation it will go out of sync.